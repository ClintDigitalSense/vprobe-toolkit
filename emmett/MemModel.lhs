Copyright 2008, VMware, Inc.

A MemModel is a pointer width and access mode. E.g., 32-bit guest pointers,
64-bit guest pointers, etc.

> module MemModel(MemModel(), nameToMemModel, defaultMemModel, pointerWidth,
>                 defaultAlign, emitRef) where
> data MemModel = Mem Int String
> instance Show MemModel where
>    show (Mem i s) = (show i) ++ (" bytes, ") ++ s
> instance Eq MemModel where
>     Mem i s == Mem j t = i == j && s == t
> nameToMemModelMap = [ ("guest32", memModelGuest32),
>                       ("guest64", memModelGuest64),
>                       ("guestphys", memModelGuestPhys),
>                       ("vmm32",   memModelVMM32),
>                       ("vmm64",   memModelVMM64),
>                       ("vmk",     memModelVMKernel) ]
> nameToMemModel s = case lookup s nameToMemModelMap of
>                      Just m -> m
>                      Nothing -> error ("unknown memory model: " ++ s)
> defaultMemModel  = memModelGuest64
> memModelVMM32 = Mem 4   "getmonitor"
> memModelVMM64 = Mem 8   "getmonitor"
> memModelGuest32 = Mem 4 "getguest"
> memModelGuest64 = Mem 8 "getguest"
> memModelGuestPhys = Mem 8 "getguestphys"
> memModelVMKernel = Mem 8 "getvmkernel"
> pointerWidth (Mem w _) = w
> defaultAlign = pointerWidth
> emitRef (Mem w s) = s ++ " "
