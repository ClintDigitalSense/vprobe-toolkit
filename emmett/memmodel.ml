(* **********************************************************
 * Copyright 2009 VMware, Inc.  All rights reserved.
 * -- VMware Confidential
 * **********************************************************)

(*
 * Memory model module: a memory model describes memory layout, i.e.,
 * alignment and width, for int, long, and pointer types. Only integer
 * and pointer types are tagged with a memory model. The layout of
 * structs, unions, and arrays follows from the the layout of their
 * inner integer and pointer components. For pointers, the memory
 * model additionally indicates the accessor function, used for
 * accessing the pointer's target memory.
 *)

type memmodel = { acc: string; i: int; l: int; p: int }
type mmident  = string 

let memModelMap = [
   "vmw32",     { acc="getvmw";       i=4; l=4; p=4 };
   "vmw64",     { acc="getvmw";       i=4; l=8; p=8 };
   "guest16",   { acc="getguest";     i=2; l=4; p=2 };
   "guest32",   { acc="getguest";     i=4; l=4; p=4 };
   "guest64",   { acc="getguest";     i=4; l=8; p=8 };
   "guestphys", { acc="getguestphys"; i=8; l=8; p=8 };
]

let mmValidMemModel(id: mmident) : bool =
    List.mem_assoc id memModelMap

let mmGetIntSize(id: mmident) : int =
    (List.assoc id memModelMap).i

let mmGetLongSize(id: mmident) : int =
    (List.assoc id memModelMap).l

let mmGetPtrSize(id: mmident) : int =
    (List.assoc id memModelMap).p

let mmGetAccessor(id: mmident) : string =
    (List.assoc id memModelMap).acc

let mmGetSize(id: mmident) (s: string) : int =
  match s with
  | "char"  -> 1
  | "short" -> 2
  | "int"   -> mmGetIntSize(id)
  | "long"  -> mmGetLongSize(id)
  | "ptr"   -> mmGetPtrSize(id)
  | _ -> failwith ("Invalid mem model size request: " ^ s)
