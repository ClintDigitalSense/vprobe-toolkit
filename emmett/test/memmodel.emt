memmodel guest64 {
    typedef struct {
        uint32  field1;
        uint32* ptr;
    } g64;
    g64 *gptr;
}

memmodel vmm32 {
    typedef struct {
        uint32  field1;
        uint32* ptr;
    } v32;
    v32 *vptr;
}

memmodel vmk {
    typedef struct {
        uint32  field1;
        uint32* ptr;
    } k64;
    k64 *kptr;
}

VMM1Hz {
   myint = *vptr->ptr + *gptr->ptr + *kptr->ptr;
}
