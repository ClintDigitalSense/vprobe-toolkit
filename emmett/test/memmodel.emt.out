(version 0.2)
(definteger gptr)
(definteger kptr)
(definteger myint)
(definteger vptr)

 ;;;;;
(vprobe VMM1Hz ( do  ( setint myint    (+  (+ (& 4294967295 (getmonitor (& 4294967295 (getmonitor  (+ vptr  4 )))))  (& 4294967295 (getguest (getguest  (+ gptr  8 )))) )  (& 4294967295 (getvmkernel (getvmkernel  (+ kptr  8 )))) ))))